perfschema/insert.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package perfschema
+
+import (
+	"strings"
+	"sync/atomic"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// InsertValues is the rest part of insert/replace into statement.
+type InsertValues struct {
+	ColNames   []string
+	Lists      [][]expression.Expression
+	TableIdent table.Ident
+	Setlist    []*expression.Assignment
+}
+
+// Simulate the behavior of an INSERT statement.
+func (ps *perfSchema) ExecInsert(insertVals *InsertValues) error {
+	t := ps.getTable(insertVals.TableIdent)
+	if t == nil {
+		return errors.Errorf("INSERT INTO %s: operation not permitted", insertVals.TableIdent)
+	}
+
+	cols, err := ps.getColumns(insertVals, t.Columns)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rows, err := ps.getRows(insertVals, t, cols)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return ps.addRecords(insertVals.TableIdent, rows)
+}
+
+func (ps *perfSchema) addRecords(tableIdent table.Ident, rows [][]interface{}) error {
+	var store *leveldb.DB
+	var lastLsn uint64
+
+	// Same as MySQL, we only support INSERT operations for setup_actors & setup_objects.
+	switch strings.ToUpper(tableIdent.Name.O) {
+	case TableSetupActors:
+		store = ps.stores[TableSetupActors]
```

Comment:
How about using a variable like `name := strings.ToUpper(tableIdent.Name.O)`, then u can directly using it.